lib/iris/_concatenate.py

```diff
@@ -281,14 +288,244 @@ class _CoordExtent(namedtuple("CoordExtent", ["points", "bounds"])):
+            chunk=_hash_chunk,
+            combine=_hash_chunk,
+            aggregate=_hash_aggregate,
+            keepdims=False,
+            meta=np.empty(tuple(), dtype=np.int64),
+            dtype=np.int64,
+        )
+    else:
+        result = _hash_aggregate(a, None, False)
+    return result
+
+
+class _ArrayHash(namedtuple("ArrayHash", ["value", "chunks"])):
+    """Container for a hash value and the chunks used when computing it.
+
+    Parameters
+    ----------
+    value : :class:`np.int64`
+        The hash value.
+    chunks : tuple
+        The chunks the array had when the hash was computed.
+    """
+
+    __slots__ = ()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            raise TypeError(f"Unable to compare {repr(self)} to {repr(other)}")
+        if self.chunks != other.chunks:
+            raise ValueError(
+                "Unable to compare arrays with different chunks: "
+                f"{self.chunks} != {other.chunks}"
+            )
+        return self.value == other.value
+
+
+def _array_id(
+    coord: DimCoord | AuxCoord | AncillaryVariable | CellMeasure,
+    bound: bool,
+) -> str:
+    """Get a unique key for looking up arrays associated with coordinates."""
+    return f"{id(coord)}{bound}"
+
+
+def _compute_hashes(
+    arrays: Mapping[str, np.ndarray | da.Array],
+) -> dict[str, _ArrayHash]:
+    """Compute hashes for the arrays that will be compared.
+
+    Two arrays are considered equal if each unmasked element compares equal
+    and the masks are equal. However, hashes depend on chunking and dtype.
+    Therefore, arrays with the same shape are rechunked so they have the same
+    chunks and arrays with numerical dtypes are cast up to the same dtype before
+    computing the hashes.
+
+    Parameters
+    ----------
+    arrays :
+        A mapping with key-array pairs.
+
+    Returns
+    -------
+    dict[str, _ArrayHash] :
+        An dictionary of hashes.
+
+    """
+    hashes = {}
+
+    def is_numerical(dtype):
+        return np.issubdtype(dtype, np.bool_) or np.issubdtype(dtype, np.number)
+
+    def group_key(item):
+        array_id, a = item
+        if is_numerical(a.dtype):
+            dtype = "numerical"
+        else:
+            dtype = str(a.dtype)
+        return a.shape, dtype
+
+    sorted_arrays = sorted(arrays.items(), key=group_key)
+    for _, group_iter in itertools.groupby(sorted_arrays, key=group_key):
+        array_ids, group = zip(*group_iter)
+        # Unify dtype for numerical arrays, as the hash depends on it
+        if is_numerical(group[0].dtype):
+            dtype = np.result_type(*group)
+            same_dtype_arrays = tuple(a.astype(dtype) for a in group)
+        else:
+            same_dtype_arrays = group
+        if any(isinstance(a, da.Array) for a in same_dtype_arrays):
+            # Unify chunks as the hash depends on the chunks.
+            indices = tuple(range(group[0].ndim))[::-1]
```

Comment:
What's the significance of reversing the order of the tuple?

Reply:
I don't quite remember, it may be a copy/paste from e.g. https://github.com/dask/dask/blob/80b0737c9ad5848fa779efb8f841b9fa35c8ff36/dask/array/overlap.py#L692-L696. Since we're not using broadcasting here and all tests are passing fine without it, I'll remove it.

Reply:
Removed in 073b25e